File: Java/src/main/java/com/gildedrose/ItemFactory.java
package com.gildedrose;

/**
 *
 * @author markd
 */
public final class ItemFactory {
    
    private static final String CHEESE_NAME = "Aged Brie";
    private static final String LEGENDARY_NAME = "Sulfuras, Hand of Ragnaros";
    private static final String TICKET_NAME = "Backstage passes to a TAFKAL80ETC concert";
    private static final String CONJURED_FLAG = "Conjured";
    
    public static GenericItem classifyItem(Item item){
        return newItem(item.name, item.sellIn, item.quality);
    }
    
    public static GenericItem newItem(String name, int sellIn, int quality){
        if(name.equals(CHEESE_NAME)){
            return newCheeseItem(name, sellIn, quality);
        }
        else if(name.equals(LEGENDARY_NAME)){
            return newLegendaryItem(name, sellIn, quality);
        }
        else if(name.equals(TICKET_NAME)){
            return newTicketItem(name, sellIn, quality);
        }
        else if(name.contains(CONJURED_FLAG)){
            return newConjuredItem(name, sellIn, quality);
        }
        else{
            return newGenericItem(name, sellIn, quality);
        }
    }
    
    public static GenericItem newGenericItem(String name, int sellIn, int quality){
        return new GenericItem(name, sellIn, quality);
    }
    
    public static CheeseItem newCheeseItem(String name, int sellIn, int quality){
        return new CheeseItem(name, sellIn, quality);
    }
    
    public static LegendaryItem newLegendaryItem(String name, int sellIn, int quality){
        return new LegendaryItem(name, sellIn, quality);
    }
    
    public static TicketItem newTicketItem(String name, int sellIn, int quality){
        return new TicketItem(name, sellIn, quality);
    }
    
    public static ConjuredItem newConjuredItem(String name, int sellIn, int quality){
        return new ConjuredItem(name, sellIn, quality);
    }
}

File: ruby/item_factory.rb
class ItemFactory

  CHEESE_NAME = 'Aged Brie'.freeze
  LEGENDARY_NAME = 'Sulfuras, Hand of Ragnaros'.freeze
  TICKET_NAME = 'Backstage passes to a TAFKAL80ETC concert'.freeze
  CONJURED_FLAG = 'Conjured'.freeze

  def self.classify_item(item)
    new_item(item.name, item.sell_in, item.quality)
  end

  def self.new_item(name, sell_in, quality)
    if name.eql?(CHEESE_NAME)
      CheeseItem.new(Item.new(name, sell_in, quality))
    elsif name.eql?(LEGENDARY_NAME)
      LegendaryItem.new(Item.new(name, 0, quality))
    elsif name.eql?(TICKET_NAME)
      TicketItem.new(Item.new(name, sell_in, quality))
    elsif name.include?(CONJURED_FLAG)
      ConjuredItem.new(Item.new(name, sell_in, quality))
    else
      GenericItem.new(Item.new(name, sell_in, quality))
    end
  end
end